Lomiri shell package, modules & tests: init

OPNA2608 opened the pull request:
## Description of changes

Working towards #99090.

Inits:

- [`lomiri.lomiri`](https://gitlab.com/ubports/development/core/lomiri), the [Lomiri Shell](https://lomiri.com/)
- [`lomiri.lomiri-session`](https://gitlab.com/ubports/development/core/lomiri-session), scripts & data files for starting a Lomiri session
- A module for enabling the Lomiri DE
- A test to check if
  - Lomiri launches
  - Basic app expectations work
  - Basic inter-app behaviour works
  - Ayatana Indicators are usable

And various fixes to Lomiri components, now that a test can be run for them.

<details>
  <summary>Some pictures, after installing it on spare hardware</summary>

![screenshot20240411_134047572](https://github.com/NixOS/nixpkgs/assets/23431373/f16117c6-e8fe-41e8-a7c3-db194949043a)
![screenshot20240411_163943601](https://github.com/NixOS/nixpkgs/assets/23431373/e9841514-e2ee-4427-9866-9dac56766d96)

</details>

Checklist:
- [x] Submit some patches upstream
- [x] Decide if we should disable the first time wizard in general (i'm not sure if it's of much use for us)
  - Decision: Will just patch it out, it mostly just asks the stuff that should be set up by the system config already.
- [x] Decide where the nixos module should be put (there might be an ongoing effort to start moving those outside of `services.xserver`: https://github.com/NixOS/nixpkgs/pull/292141)
  - ~~Decision: For now, `services.xserver.desktopManager.lomiri` is prolly fine. If there's any developments on the linked PR, then I might change it to `services.desktopManager.lomiri` (although we'd also need a display technology agnostic `services.displayManager.lightdm` for the Wayland `services.displayManager.lightdm.greeters.lomiri`).~~
  - Decision: With #291913 merged, names are now `services.desktopManager.lomiri` (no longer implying X11) and `services.xserver.displayManager.lightdm.greeter.lomiri` (`lightdm` is yet to be moved out of `services.xserver`)
- [x] Clean up the module, check what's really necessary, resolve leftover errors
  - Left some TODOs for myself when further stuff gets upstreamed, rest looks fine to me.
- [x] Apparently, Debian got the lomiri greeter working? Check if we can get that working too
  - Got the greeter working
- [x] Release notes entry about Lomiri availability
  - Entry has been written, sprinkling in some details about what it & our state of it is. Not sure if it's too much, happy about feedback on it.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

OPNA2608 committed:
lomiri.lomiri: init at 0.2.1

OPNA2608 committed:
lomiri.lomiri-session: init at 0.2

OPNA2608 committed:
nixos/lomiri: init

OPNA2608 committed:
nixos/tests/lomiri: init

OPNA2608 committed:
lomiri.morph-browser: Patch desktop icon path to be absolute

Because Lomiri App Launch parses it on Lomiri startup, and having a relative Icon entry somehow corrupts the search paths.

OPNA2608 committed:
lomiri.content-hub: Move example peers to examples output

OPNA2608 committed:
lomiri.lomiri-system-settings: Patch language plugin's locale lookup path

The only part this really affects for us is Lomiri's First-Time-Launch Wizard, which uses the found & filtered
locale identifiers for a language selection & gets stuck unless at least 1 valid language has been found.

This makes the wizard process completable, in case we ever re-enable it.

OPNA2608 committed:
nixos/lightdm-greeters/lomiri: init

SuperSandro2000 commented:
> * Decision: For now, `services.xserver.desktopManager.lomiri` is prolly fine. If there's any developments on the linked PR, then I might change it to `services.desktopManager.lomiri` (although we'd also need a display technology agnostic `services.displayManager.lightdm` for the Wayland `services.displayManager.lightdm.greeters.lomiri`).
> *

FYI https://github.com/NixOS/nixpkgs/pull/291913

SuperSandro2000 commented on nixos/modules/programs/lomiri.nix:
```diff
@@ -0,0 +1,160 @@
+  cfg = config.programs.lomiri;
+in {
+  options.programs.lomiri= {
+    enable = lib.mkEnableOption (lib.mdDoc ''
+      the Lomiri graphical shell (formerly known as Unity8)'');
```
I am not sure if adding a line break here breaks the docs.

SuperSandro2000 commented on nixos/modules/programs/lomiri.nix:
```diff
@@ -0,0 +1,160 @@
+
+    # Copy-pasted
+    # TODO are all of these needed? just nice-have's? convenience?
+    hardware.opengl.enable = lib.mkDefault true;
+    fonts.enableDefaultPackages = lib.mkDefault true;
+    programs.dconf.enable = lib.mkDefault true;
+    programs.xwayland.enable = lib.mkDefault true;
```
I think those should become part of a shared module between xserver and wayland envs. After splitting out xserver this is on my todo

OPNA2608 replied:
Agree, will keep them copy-pasted here until then.

SuperSandro2000 commented on nixos/modules/programs/lomiri.nix:
```diff
@@ -0,0 +1,160 @@
+
+    services.telepathy.enable = true;
+
+    services.xserver.displayManager.lightdm.enable = lib.mkDefault true;
+    services.xserver.displayManager.lightdm.greeters.lomiri.enable = lib.mkDefault true;
+    services.xserver.displayManager.defaultSession = lib.mkDefault "lomiri";
+    services.xserver.displayManager.sessionPackages = with pkgs.lomiri; [ lomiri-session ];
```
```suggestion
    services.xserver.displayManager = {
      lightdm.enable = lib.mkDefault true;
      lightdm.greeters.lomiri.enable = lib.mkDefault true;
      defaultSession = lib.mkDefault "lomiri";
      sessionPackages = with pkgs.lomiri; [ lomiri-session ];
    };
```

OPNA2608 replied:
For consistency:
```suggestion
    services.xserver.displayManager = {
      lightdm = {
        enable = lib.mkDefault true;
        greeters.lomiri.enable = lib.mkDefault true;
      };
      defaultSession = lib.mkDefault "lomiri";
      sessionPackages = with pkgs.lomiri; [ lomiri-session ];
    };
```

SuperSandro2000 commented on nixos/modules/programs/lomiri.nix:
```diff
@@ -0,0 +1,160 @@
+      };
+    };
+
+    systemd.tmpfiles.rules = [
```
Please use the new tmpfiles syntax

OPNA2608 replied:
Please add references when you say stuff like this. If I was already aware of what the "new syntax" was then I wouldn't be submitting it like this.

The syntax of `systemd.tmpfiles.rules` itself didn't change, so I'm *assuming* you're referring to (linking these here for when I have time to rewrite it):

- new option for specifying tmpfile rules: https://github.com/NixOS/nixpkgs/pull/262731
- example migration to new API: https://github.com/NixOS/nixpkgs/pull/280373

OPNA2608 replied:
Avoided the tmpfiles usage here via other means, but I adjusted the one in the tests to use `tmpfiles.settings` instead.

SuperSandro2000 reviewed (commented):
Already looking pretty solid

OPNA2608 commented on pkgs/desktops/lomiri/applications/lomiri/default.nix:
```diff
@@ -0,0 +1,288 @@
+
+    # Use Nix flake instead of Canonical's Ubuntu logo
+    rm qml/Launcher/graphics/home.svg
+    ln -s ${./lomiri-launcher-logo.svg} qml/Launcher/graphics/home.svg
```
When I last fiddled with this ~ a year ago, using regular snowflake svg didn't work correctly / looked broken in this context. I'll take another stab at it.

OPNA2608 replied:
Ah right. Using `${nixos-icons}/share/icons/hicolor/scalable/apps/nix-snowflake-white.svg` *works*, but the icon is displayed abit squished.

![image](https://github.com/NixOS/nixpkgs/assets/23431373/2b79c950-9a6c-4a7e-97f8-bb0c9c0b92f2)

This is because none of the snowflake graphics have correct dimensions. They should all have widths that are equal to their heights & the XDG icon directories they're in, but none of them have that.

https://github.com/NixOS/nixos-artwork/issues/42

```
/nix/store/9v1j72wdckazf1301lag60kapx0zcxcv-nixos-icons-2021-02-24/share/icons/hicolor/32x32/apps/nix-snowflake-white.png PNG 32x28 32x28+0+0 8-bit Grayscale Gray 929B 0.000u 0:00.000
/nix/store/9v1j72wdckazf1301lag60kapx0zcxcv-nixos-icons-2021-02-24/share/icons/hicolor/128x128/apps/nix-snowflake-white.png PNG 128x111 128x111+0+0 8-bit Grayscale Gray 4159B 0.000u 0:00.000
/nix/store/9v1j72wdckazf1301lag60kapx0zcxcv-nixos-icons-2021-02-24/share/icons/hicolor/scalable/apps/nix-snowflake-white.svg SVG 535x464 535x464+0+0 16-bit sRGB 6695B 0.000u 0:00.007
```

I guess having it squished would be fine, but it just really bothered me to look at that all the time…

OPNA2608 replied:
Fixing this upstream: https://github.com/NixOS/nixos-artwork/pull/118

I guess we'll use the not-quite-square ones until then, so the fix propagates into here eventually…

OPNA2608 replied:
Fixed upstream, updating `nixos-icons` in https://github.com/NixOS/nixpkgs/pull/303068 . Will switch now, and let that propagate via the update.

OPNA2608 commented on pkgs/desktops/lomiri/applications/morph-browser/default.nix:
```diff
@@ -56,9 +56,12 @@ stdenv.mkDerivation (finalAttrs: {
+    # As a compromise, hardcode /run/current-system
     substituteInPlace src/app/webbrowser/morph-browser.desktop.in.in \
-      --replace 'Icon=@CMAKE_INSTALL_FULL_DATADIR@/morph-browser/morph-browser.svg' 'Icon=morph-browser' \
+      --replace 'Icon=@CMAKE_INSTALL_FULL_DATADIR@/morph-browser/morph-browser.svg' 'Icon=/run/current-system/sw/share/icons/hicolor/scalable/apps/morph-browser.svg' \
```
I was told that we don't want to refer to `$out` in desktop files, because some environments may copy them to separate locations and the path references would become stale. I.e., I think Pantheon made a `~/.config/autostart/` with copies of various desktop files.

wineee commented on nixos/modules/programs/lomiri.nix:
```diff
@@ -0,0 +1,160 @@
+    services.xserver.displayManager.lightdm.greeters.lomiri.enable = lib.mkDefault true;
+    services.xserver.displayManager.defaultSession = lib.mkDefault "lomiri";
+    services.xserver.displayManager.sessionPackages = with pkgs.lomiri; [ lomiri-session ];
+    services.xserver.updateDbusEnvironment = true;
```
`updateDbusEnvironment` don't work for session configured by `services.xserver.displayManager.sessionPackages`, 

It add  `${lib.getBin pkgs.dbus}/bin/dbus-update-activation-environment --systemd --all ` to `xsession` script, which only be generated when  `services.xserver.displayManager.session` was used

https://github.com/NixOS/nixpkgs/issues/209847

OPNA2608 replied:
Will drop it from here.

OPNA2608 committed:
lomiri.libusermetrics: Add custom patch for launching systemd service

The UserMetrics service expects AppArmor to be available, and its database access breaks when that's not the case.
Details: https://gitlab.com/ubports/development/core/libusermetrics/-/issues/8

We need to set an envvar for it to work AppArmor-less, but that requires system config knowledge.
Solve this by telling the D-Bus service to look for & launch a systemd service, which we will later create in the Lomiri module.

OPNA2608 committed:
lomiri.telephony-service: Fix indicator

OPNA2608 committed:
ayatana-indicator-messages: Fix desktop parsing

Not *great*, but a start.

MessagingMenu has a function for mapping an app to desktop file data. This relies on GLib to parse the desktop file,
which needs the binary from the 'Exec=' entry on PATH, otherwise it bails.

messaging_menu_app_set_desktop_id calls g_desktop_app_info_new:
https://github.com/AyatanaIndicators/ayatana-indicator-messages/blob/334668ad5ad33c6bd342f34b159dbcbed8c65df1/libmessaging-menu/messaging-menu-app.c#L245-L252

Which, after some jumps inbetween, ends up in g_desktop_app_info_load_from_keyfile and calls g_find_program_for_path:
https://github.com/GNOME/glib/blob/204a76669bd46c5eb656946e21bd746685c67174/gio/gdesktopappinfo.c#L1920-L1951

Which uses PATH:
https://github.com/GNOME/glib/blob/204a76669bd46c5eb656946e21bd746685c67174/glib/gutils.c#L371

OPNA2608 commented:
@ofborg build nixosTests.morph-browser nixosTests.lomiri

OPNA2608 committed:
doc/release-notes: mention Lomiri availability

OPNA2608 commented:
*Hoping* this is ready now.

SuperSandro2000 reviewed (approved):
Didn't try myself but LGTM

OPNA2608 commented on nixos/modules/services/x11/display-managers/lightdm-greeters/lomiri.nix:
```diff
@@ -0,0 +1,34 @@
+    };
+  };
+
+  config = lib.mkIf (ldmcfg.enable && cfg.enable) {
```
Note to self: I think this module is highly likely to be missing some stuff in case `services.xserver.displayManager.lightdm.greeters.lomiri.enable && (!services.desktopManager.lomiri.enable)`, basically requiring some code duplication (since the shell is also the greeter).

IMO that's *prolly* not worth blocking this on? I'm fine with a DE-specific DM being abit broken when not using said DE, will just put it on my TODO list.

wegank commented on nixos/modules/services/desktop-managers/lomiri.nix:
```diff
@@ -0,0 +1,165 @@
+  cfg = config.services.desktopManager.lomiri;
+in {
+  options.services.desktopManager.lomiri= {
+    enable = lib.mkEnableOption (lib.mdDoc ''
```
`lib.mdDoc` was removed on master.

OPNA2608 replied:
Ah right, I saw https://github.com/NixOS/nixpkgs/pull/303841 yesterday but forgot to subscribe to it & address the removal here. Thanks!
